File: src/main/java/sage/web/RelationController.java
package sage.web;

import java.util.Collection;
import java.util.Collections;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import sage.domain.service.HeedService;
import sage.domain.service.NotifService;
import sage.domain.service.RelationService;
import sage.web.auth.Auth;

@RestController
@RequestMapping(method = RequestMethod.POST)
public class RelationController {
  @Autowired
  private RelationService relationService;
  @Autowired
  private NotifService notifService;
  @Autowired
  private HeedService heedService;

  @RequestMapping("/follow/{targetId}")
  public void follow(@PathVariable Long targetId,
      @RequestParam(required = false) String reason,
      @RequestParam(value = "tagIds[]", required = false) Collection<Long> tagIds,
      @RequestParam(required = false) Boolean includeNew,
      @RequestParam(required = false) Boolean includeAll) {
    Long cuid = Auth.checkCuid();

    if (tagIds == null) {
      tagIds = Collections.emptyList();
    }
    relationService.follow(cuid, targetId, reason, tagIds, boolValue(includeNew), boolValue(includeAll));
    // Send "followed" notification
    notifService.followed(targetId, cuid);
  }

  @RequestMapping("/unfollow/{targetId}")
  public void unfollow(@PathVariable Long targetId) {
    Long uid = Auth.checkCuid();

    relationService.unfollow(uid, targetId);
  }

  private boolean boolValue(Boolean o) {
    return o != null ? o : false;
  }
}

File: src/main/kotlin/sage/web/page/RelationController.kt
package sage.web.page

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.servlet.ModelAndView
import sage.service.RelationService
import sage.service.UserService
import sage.web.auth.Auth

@Controller
class RelationController @Autowired constructor(
    private val userService: UserService,
    private val relationService: RelationService) {

  @RequestMapping("/followings")
  fun followings() = "forward:/followings/${Auth.checkUid()}"

  @RequestMapping("/followings/{userId}")
  fun followings(@PathVariable userId: Long): ModelAndView {
    val uid = Auth.checkUid()
    val thisUser = userService.getUserCard(uid, userId)
    val followings = relationService.followings(userId).map { fol ->
      userService.getUserCard(uid, fol.target.id)
    }
    return ModelAndView("followings").addObject("users", followings)
        .addObject("thisUser", thisUser)
}

  @RequestMapping("/followers")
  fun followers() = "forward:/followers/${Auth.checkUid()}"

  @RequestMapping("/followers/{userId}")
  fun followers(@PathVariable userId: Long): ModelAndView {
    val uid = Auth.checkUid()
    val thisUser = userService.getUserCard(uid, userId)
    val followers = relationService.followers(userId).map { fol ->
      userService.getUserCard(uid, fol.source.id)
    }
    return ModelAndView("followers").addObject("users", followers)
        .addObject("thisUser", thisUser)
  }
}
